File: dsa-pros/leetcode/array/JavaScript/timiBello/Week 4/wk4.py
#Contains Duplicate

class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        seen = set()
        for num in nums:
            if num in seen:
                return True
            seen.add(num)
        return False


#Binary Search
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            middle = (left + right) // 2
            if nums[middle] == target:
                return middle
            elif nums[middle] > target:
                right = middle - 1
            else:
                    left = middle + 1
        return -1


#Rotate Array
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        k = k % len(nums)  # Effective number of rotations
        nums[:] = nums[-k:] + nums[:-k]


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead. Solution to do it in-place with 0(1) extra space. 
        """
        k = k % len(nums)  # Effective number of rotations
        count = 0  # Number of elements rotated
        start = 0
        while count < len(nums):
            current = start
            prev = nums[start]
            
            while True:
                next_idx = (current + k) % len(nums)
                nums[next_idx], prev = prev, nums[next_idx]
                current = next_idx
                count += 1
                
                if start == current:
                    break
                    
            start += 1


#Longest Increasing Subsequence
#  On square time complexity using dp approach
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0 
        
        n = len(nums)
        dp = [1] * n  # Initialize dp array with all elements as 1
        
        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)

        return max(dp)

#Finding it tough to come up with an algorithm that runs in O(n log(n)) time complexity that gives the expected output. 

File: dsa-pros/leetcode/array/JavaScript/timiBello/Week 4/wk4.ts
/*Contains Duplicate*/
function containsDuplicate(nums) {
  const seen = new Set();
  for (let num of nums) {
    if (seen.has(num)) {
      return true;
    }
    seen.add(num);
  }
  return false;
}

//Binary Search
function search(nums: number[], target: number): number {
let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    let middle = Math.floor((left + right) / 2);

    if (nums[middle] === target) {
      return middle;
    } else if (nums[middle] > target) {
      right = middle - 1;
    } else {
      left = middle + 1;
    }
  }

  return -1;
};

/* Rotate Array
Do not return anything, modify nums in-place instead.*/
function rotate(nums: number[], k: number): void {
      k = k % nums.length; // Effective number of rotations
  let count = 0; // Number of elements rotated

  let start = 0;
  while (count < nums.length) {
    let current = start;
    let prev = nums[start];

    do {
      const nextIdx = (current + k) % nums.length;
      [nums[nextIdx], prev] = [prev, nums[nextIdx]];
      current = nextIdx;
      count++;
    } while (start !== current);

    start++;
  }
};


/*Longest Increasing Subsequence
(did this to give 0(n log(n)) time complexity using Dynamic Programming with Binary Search algorithm*/
function lengthOfLIS(nums: number[]): number {
if (nums.length === 0) {
        return 0;
    }

    const tails = [nums[0]]; // Initialize the tails array with the first element

    for (let i = 1; i < nums.length; i++) {
        const num = nums[i];

        if (num > tails[tails.length - 1]) {
            tails.push(num);
        } else {
            // Use binary search to find the correct position to update the tail
            let left = 0;
            let right = tails.length - 1;

            while (left < right) {
                const mid = Math.floor((left + right) / 2);

                if (tails[mid] < num) {
									 left = mid + 1;
                } else {
                    right = mid;
                }
            }

            tails[left] = num;
        }
    }

    return tails.length;
};
